x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/support/KerberosTicketValidator.java

```diff
@@ -0,0 +1,271 @@
+    private static final String SUN_KRB5_LOGIN_MODULE = "com.sun.security.auth.module.Krb5LoginModule";
+
+    /**
+     * Validates client kerberos ticket received from the peer.
+     * <p>
+     * First performs service login using keytab, supports multiple principals in
+     * keytab and the principal is selected based on the request.
+     * <p>
+     * The GSS security context establishment is handled and if it is established then
+     * returns the Tuple of username and out token for peer reply. It may return
+     * Tuple with null username but with a outToken to be sent to peer for further
+     * negotiation.
+     *
+     * @param decodedToken base64 decoded kerberos ticket bytes
+     * @param keytabPath Path to Service key tab file containing credentials for ES
+     *            service.
+     * @param krbDebug if {@code true} enables jaas krb5 login module debug logs.
+     * @return {@link Tuple} of user name {@link GSSContext#getSrcName()} and out
+     *         token base64 encoded if any. When context is not yet established user
+     *         name is {@code null}.
+     * @throws LoginException thrown when service authentication fails
+     *             {@link LoginContext#login()}
+     * @throws GSSException thrown when GSS Context negotiation fails
+     *             {@link GSSException}
+     */
+    public Tuple<String, String> validateTicket(final byte[] decodedToken, final Path keytabPath, final boolean krbDebug)
+            throws LoginException, GSSException {
+        final GSSManager gssManager = GSSManager.getInstance();
+        GSSContext gssContext = null;
+        LoginContext loginContext = null;
+        try {
+            loginContext = serviceLogin(keytabPath.toString(), krbDebug);
+            GSSCredential serviceCreds = createCredentials(gssManager, loginContext.getSubject());
+            gssContext = gssManager.createContext(serviceCreds);
+            final String base64OutToken = base64Encode(acceptSecContext(decodedToken, gssContext, loginContext.getSubject()));
+            LOGGER.trace("validateTicket isGSSContextEstablished = {}, username = {}, outToken = {}", gssContext.isEstablished(),
+                    gssContext.getSrcName().toString(), base64OutToken);
+            return new Tuple<>(gssContext.isEstablished() ? gssContext.getSrcName().toString() : null, base64OutToken);
+        } catch (PrivilegedActionException pve) {
+            if (pve.getCause() instanceof LoginException) {
+                throw (LoginException) pve.getCause();
+            }
+            if (pve.getCause() instanceof GSSException) {
+                throw (GSSException) pve.getCause();
+            }
+            throw ExceptionsHelper.convertToRuntime(pve.getException());
+        } finally {
+            privilegedLogoutNoThrow(loginContext);
+            privilegedDisposeNoThrow(gssContext);
+        }
+    }
+
+    private String base64Encode(final byte[] outToken) {
```

Comment:
can you rename this method? This has specific behavior for an outgoing token; ie check for null/empty